### PR TITLE
adding storage layer to store & retrieve & delete scan requests

### PR DESCRIPTION
**Pull request by devaradhanm**
#### Description of changes
- added support to delete documents after each test for db tests
- added converter to transform between view model & storage document. (not used yet)

#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Added relevant unit test for your changes. (`yarn test`)
- [ ] Verified code coverage for the changes made. Check coverage report at: `<rootDir>/test-results/unit/coverage`
- [ ] Ran precheckin (`yarn precheckin`)

### CODE REVIEW

File: packages/service-library/src/models/on-demand-page-scan-result-model.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
```
Comment (devaradhanm) [inline packages/service-library/src/models/on-demand-page-scan-result-model.ts]: i have just wrapped the code in to a class & added a new method deleteAllDocuments

File: packages/azure-services/src/ioc-types.ts [hunk 2]
```diff
@@ -24,4 +24,6 @@ export type QueueServiceURLProvider = () => Promise<ServiceURL>;
 
 export const cosmosContainerClientTypes = {
     A11yIssuesCosmosContainerClient: 'a11yIssuesCosmosContainerClient',
+    UnProcessedPageScanRequestsCosmosContainerClient: 'unprocessedPageScanRequestsCosmosContainerClient',
```
Comment (lamaks) [inline packages/azure-services/src/ioc-types.ts]: Unprocessed is one word

File: packages/service-library/src/models/unprocessed-page-request-model.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+export interface UnProcessedPageScanRequestModel {
```
Comment (lamaks) [inline packages/service-library/src/models/unprocessed-page-request-model.ts]: one word

File: packages/service-library/src/converters/converter.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+export interface Converter<DocumentType, ViewModelType> {
```
Comment (lamaks) [inline packages/service-library/src/converters/converter.ts]: Can we then move /factories/ under /converters/ to unify or vs

File: packages/service-library/src/data-providers/document-provider.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { client, CosmosOperationResponse } from 'azure-services';
+import { injectable } from 'inversify';
+
+@injectable()
+export abstract class DocumentProvider {
+    protected async executeQueryWithContinuationToken<T>(execute: (token?: string) => Promise<CosmosOperationResponse<T[]>>): Promise<T[]> {
```
Comment (lamaks) [inline packages/service-library/src/data-providers/document-provider.ts]: this is duplicate code from PageDocumentProvider Can we move it under /azure-services/CosmosClientWrapper/ package to expose and refactor out duplicates

File: packages/service-library/src/data-providers/unprocessed-scan-request-provider.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { CosmosContainerClient, cosmosContainerClientTypes, CosmosOperationResponse } from 'azure-services';
+import { inject } from 'inversify';
+import { ItemType, UnProcessedPageScanRequest } from 'storage-documents';
+import { DocumentProvider } from './document-provider';
+
+export class UnProcessedScanRequestProvider extends DocumentProvider {
```
Comment (lamaks) [inline packages/service-library/src/data-providers/unprocessed-scan-request-provider.ts]: one word

File: packages/service-library/src/data-providers/unprocessed-scan-request-provider.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    public static readonly partitionKey: string = 'unProcessedScanRequestDocuments';
```
Comment (lamaks) [inline packages/service-library/src/data-providers/unprocessed-scan-request-provider.ts]: one word

File: packages/storage-documents/src/unprocessed-page-scan-request.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { StorageDocument } from '.';
+import { ItemType } from './item-type';
+
+export interface UnProcessedPageScanRequest extends StorageDocument {
```
Comment (lamaks) [inline packages/storage-documents/src/unprocessed-page-scan-request.ts]: Can we call it OnDemandPageScanRequest instead?

File: packages/service-library/src/models/on-demand-page-scan-result-model.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+import { OnDemandPageScanReport, OnDemandPageScanRunResult, OnDemandScanResult } from 'storage-documents';
+
```
Comment (lamaks) [inline packages/service-library/src/models/on-demand-page-scan-result-model.ts]: -model.ts in a file name is redundant

File: packages/azure-services/src/register-azure-services-to-container.ts [hunk 10]
```diff
@@ -51,6 +51,14 @@ export function registerAzureServicesToContainer(container: Container): void {
         return createCosmosContainerClient(context.container, 'scanner', 'a11yIssues');
     });
 
+    container.bind(cosmosContainerClientTypes.UnProcessedPageScanRequestsCosmosContainerClient).toDynamicValue(context => {
+        return createCosmosContainerClient(context.container, 'scanner', 'unprocessedScanRequests');
```
Comment (lamaks) [inline packages/azure-services/src/register-azure-services-to-container.ts]: similar can we rename it to just scanRequests

File: packages/service-library/src/converters/converter.ts (hunk 4 above)
Comment (devaradhanm) [reply packages/service-library/src/converters/converter.ts]: factories are not converters rite?

File: packages/storage-documents/src/unprocessed-page-scan-request.ts [hunk 11]
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { StorageDocument } from '.';
+import { ItemType } from './item-type';
+
+export interface OnDemandPageScanRequest extends StorageDocument {
```
Comment (lamaks) [inline packages/storage-documents/src/unprocessed-page-scan-request.ts]: rename file to match class name

File: packages/service-library/src/data-providers/page-scan-request-provider.db.spec.ts [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+import 'reflect-metadata';
+
+import { CosmosContainerClient } from 'azure-services';
+import { Logger } from 'logger';
+import { ItemType, OnDemandPageScanRequest } from 'storage-documents';
+import { Mock } from 'typemoq';
+import { DbMockHelper } from '../test-utilities/db-mock-helpers';
+import { PageScanRequestProvider } from './page-scan-request-provider';
+
+describe('UnProcessedScanRequestProvider.Db', () => {
```
Comment (lamaks) [inline packages/service-library/src/data-providers/page-scan-request-provider.db.spec.ts]: rename to corresponding  method

File: packages/service-library/src/data-providers/page-scan-request-provider.db.spec.ts [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+            await dbHelper.deleteAllDocuments();
+        });
+
+        it('stores & retrieve scan results sorted by priority', async () => {
+            const request1: OnDemandPageScanRequest = {
+                id: 'id1',
+                url: 'url1',
+                priority: 10,
+                itemType: ItemType.OnDemandPageScanRequests,
+                partitionKey: 'unProcessedScanRequestDocuments',
```
Comment (lamaks) [inline packages/service-library/src/data-providers/page-scan-request-provider.db.spec.ts]: match to actual PK name

File: packages/service-library/src/data-providers/page-scan-request-provider.db.spec.ts [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+    it('no-op', () => {
+        // this test exists to have at least 1 test in the test suite to avoid jest failure, when db test run is not supported.
+    });
+
+    // tslint:disable-next-line: mocha-no-side-effect-code
+    if (dbHelper.isDbTestSupported()) {
+        let testSubject: PageScanRequestProvider;
+
+        beforeAll(async () => {
+            await dbHelper.init('test-db', 'unprocessed-scans');
```
Comment (lamaks) [inline packages/service-library/src/data-providers/page-scan-request-provider.db.spec.ts]: match to actual container

File: packages/storage-documents/src/item-type.ts [hunk 15]
```diff
@@ -5,5 +5,6 @@ export enum ItemType {
     page = 'page',
     pageScanResult = 'pageScanResult',
     issueScanResult = 'issueScanResult',
-    pageScanRunResult = 'pageScanRunResult',
+    onDemandPageScanRunResult = 'pageScanRunResult',
+    OnDemandPageScanRequests = 'pageScanRequests',
```
Comment (lamaks) [inline packages/storage-documents/src/item-type.ts]: lower case - it seems lint rules are broken, it is possible to have unused vars and imports now

**Pull request merged**